modules/juce_audio_devices/native/juce_android_Audio.cpp

```diff
@@ -440,6 +480,46 @@ class AndroidAudioIODevice  : public AudioIODevice,
         }
     }
 
+    void calculateAndReportOutputLatency(int64_t presentationFrame, int64_t presentationTime, int64_t nextOutputFrame, int64_t currentTime)
+    {
+        // Currently, the reported audio output latency on Android is only used for video playback.
+        // Thus, we don't need to report small changes in latency, so we keep the limit at 30 ms.
+        const float latencyChangeLimitSec = 0.03f;
+        const int64_t latencyChangeLimitInSamples = static_cast<int64_t>(outputSampleRate * latencyChangeLimitSec);
+
+        // Calculate when the next frame will be presented.
+        int64_t frameIndexDelta = nextOutputFrame - presentationFrame;
+        int64_t frameTimeDelta = (frameIndexDelta * 1000 * 1000 * 1000) / outputSampleRate;
+        int64_t nextFramePresentationTime = presentationTime + frameTimeDelta;
+        // Latency is the difference between the next frame presentation time and current time.
+        int64_t latencyNs = nextFramePresentationTime - currentTime;
+        int64_t latencySamples = latencyNs * outputSampleRate / (1000 * 1000 * 1000);
+
+        // Next we determine, whether the change is big enough to report to the relevant callback.
+        bool significantChange = false;
+        if (currentOutputLatencyInSamples == 0)
+        {
+            // First real measurement; take the value into use.
+            currentOutputLatencyInSamples = static_cast<int>(latencySamples);
+            significantChange = true;
+        }
+        else if (std::abs(latencySamples - currentOutputLatencyInSamples) > latencyChangeLimitInSamples)
+        {
+            // Change the latency to the mid point. This should reduce fluctuation that depends on scheduling.
+            currentOutputLatencyInSamples = static_cast<int>(currentOutputLatencyInSamples + latencySamples) / 2;
+            significantChange = true;
+        }
+
+        if (significantChange)
+        {
+            const ScopedLock sl (callbackLock);
+            if (callback != nullptr)
+            {
+                callback->audioDeviceOutputLatencyChanged(currentOutputLatencyInSamples);

```

Comment:
What are we doing with this callback? It's happening in the audio thread...

Reply:
We schedule an action on another thread that gets the current latency, and updates the app accordingly. The fact that it happens from the audio thread might be a good addition to the documentation.